### PR TITLE
add_subplot(..., axes_class=...) for more idiomatic mpl_toolkits usage.

### PR DESCRIPTION
**Pull request by anntzer**
## PR Summary

Implements https://github.com/matplotlib/matplotlib/issues/18548, see discussion there.

## PR Checklist

- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [ ] New features are documented, with examples if plot related.
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [ ] Conforms to Matplotlib style conventions (install `flake8-docstrings` and `pydocstyle<4` and run `flake8 --docstring-convention=all`).
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

File: examples/axes_grid1/demo_axes_divider.py [hunk 1]
```diff
@@ -33,10 +33,12 @@ def demo_locatable_axes_hard(fig):
     divider = SubplotDivider(fig, 2, 2, 2, aspect=True)
 
     # axes for image
-    ax = Axes(fig, divider.get_position())
+    ax = fig.add_axes(divider.get_position(), axes_class=Axes)
 
     # axes for colorbar
-    ax_cb = Axes(fig, divider.get_position())
+    # (the label prevents Axes.add_axes from incorrectly believing that the two
```
Comment (anntzer) [inline examples/axes_grid1/demo_axes_divider.py:39]: this is https://github.com/matplotlib/matplotlib/pull/10660.

**Review by timhoffm — APPROVED** [review]

File: lib/matplotlib/figure.py [hunk 2]
```diff
@@ -1073,30 +1073,39 @@ def fixlist(args):
...
-                    projection)
-            projection = 'polar'
-
-        if isinstance(projection, str) or projection is None:
-            projection_class = projections.get_projection_class(projection)
-        elif hasattr(projection, '_as_mpl_axes'):
-            projection_class, extra_kwargs = projection._as_mpl_axes()
-            kwargs.update(**extra_kwargs)
+                    "Cannot combine 'axes_class' and 'projection' or 'polar'")
+            projection_class = axes_class
```
Comment (jklymak) [inline lib/matplotlib/figure.py:1089]: You and I have crossing PRs, so its pretty important for my rebases that you have tests if possible

Comment (anntzer) [reply lib/matplotlib/figure.py:1089]: sorry about that, added a test

**Comment by anntzer** [comment]
now with docs!

File: lib/matplotlib/figure.py [hunk 3]
```diff
@@ -1129,6 +1138,10 @@ def add_axes(self, *args, **kwargs):
         polar : bool, default: False
             If True, equivalent to projection='polar'.
 
+        axes_class : subclass type of `~.axes.Axes`, optional
+            The `.axes.Axes` subclass that is instantiated.  This parameter
+            is incompatible with *projection* and *polar*.
```
Comment (jklymak) [inline lib/matplotlib/figure.py:1143]: This is somewhat mysterious.  Can we link to an example here?

Comment (anntzer) [reply lib/matplotlib/figure.py:1143]: do you want to pick one? :)

Comment (jklymak) [reply lib/matplotlib/figure.py:1143]: I dunno, which ever subclass gets used the most?

Comment (anntzer) [reply lib/matplotlib/figure.py]: I found the right thing to link.

File: lib/matplotlib/figure.py (hunk 2 above)
Comment (jklymak) [reply lib/matplotlib/figure.py:1089]: This line still doesn't have a test according to codecov.  I can probably be careful in my rebase, but....

Comment (anntzer) [reply lib/matplotlib/figure.py:1089]: hopefully better now

**Review by jklymak — APPROVED** [review]

**Pull request merged**